### PR TITLE
Ghost Roles

### PR DESCRIPTION
**Pull request by Zumorica**
## About the PR 
Adds a system for handling roles that can be taken by ghosts.

How to make the system work with prototypes and stuff:
- If you want to make an entity that will be available for takeover by ghosts once spawned, it needs the `GhostTakeoverAvailableComponent`.
- If you want to make a "spawner" prototype whose entity'll spawn a different prototype once taken over by ghosts, it needs the `GhostRoleMobSpawnerComponent`.

How to make the system work while in-game, for badmins:
- If you want to make an existing entity be available for takeover by ghosts, add `GhostTakeoverAvailableComponent` to it, and optionally set the name and description in VV.
- If you want to make a "spawner" entity, give any entity the `GhostRoleMobSpawnerComponent` component, and set the prototype, and number of available spots in VV.

**Screenshots**

image
image
image
image

### CODE REVIEW

File: Content.Server/GameObjects/Components/Observer/GhostRoleMobSpawnerComponent.cs [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+        [ViewVariables]
+        private int _currentTakeovers = 0;
+
+        [CanBeNull, ViewVariables(VVAccess.ReadWrite)] public string Prototype { get; private set; }
+
+        public override void ExposeData(ObjectSerializer serializer)
+        {
+            base.ExposeData(serializer);
+
+            serializer.DataField(this, x => Prototype, "prototype", null);
```
Comment (metalgearsloth) [inline Content.Server/GameObjects/Components/Observer/GhostRoleMobSpawnerComponent.cs:40]:
```suggestion
            serializer.DataField(this, x => x.Prototype, "prototype", null);
```

File: Content.Server/GameObjects/Components/Observer/GhostRoleMobSpawnerComponent.cs [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+        public override bool Take(IPlayerSession session)
+        {
+            if (Taken)
+                return false;
+
+            if(string.IsNullOrEmpty(Prototype))
+                throw new NullReferenceException("Prototype string cannot be null or empty!");
+
+            var mob = _entityMan.SpawnEntity(Prototype, Owner.Transform.Coordinates);
```
Comment (metalgearsloth) [inline Content.Server/GameObjects/Components/Observer/GhostRoleMobSpawnerComponent.cs:53]:
```suggestion
            var mob = Owner.EntityManager.SpawnEntity(Prototype, Owner.Transform.Coordinates);
```

File: Content.Server/GameObjects/Components/Observer/GhostRoleMobSpawnerComponent.cs [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+{
+    /// <summary>
+    ///     Allows a ghost to take this role, spawning a new entity.
+    /// </summary>
+    [RegisterComponent, ComponentReference(typeof(GhostRoleComponent))]
+    public class GhostRoleMobSpawnerComponent : GhostRoleComponent
+    {
+        public override string Name => "GhostRoleMobSpawner";
+
+        [Dependency] private readonly IServerEntityManager _entityMan = default!;
```
Comment (metalgearsloth) [inline Content.Server/GameObjects/Components/Observer/GhostRoleMobSpawnerComponent.cs:23]:
```suggestion
```

See below

File: Content.Server/GameObjects/EntitySystems/GhostRoleSystem.cs [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+        {
+            if (session.AttachedEntity == null || !session.AttachedEntity.HasComponent<GhostComponent>())
+                return;
+
+            if(_openUis.ContainsKey(session))
+                CloseEui(session);
+
+            var eui = _openUis[session] = new GhostRolesEui();
+            _euiManager.OpenEui(eui, session);
+            eui.StateDirty();
```
Comment (metalgearsloth) [inline Content.Server/GameObjects/EntitySystems/GhostRoleSystem.cs:52]: If a new Eui is made every time it's opened should the events in it be unsubscribed? So the ref gets GC'd.

File: Content.Server/GameObjects/EntitySystems/GhostRoleSystem.cs [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+    {
+        public string Command => "ghostroles";
+        public string Description => "Opens the ghost role request window.";
+        public string Help => $"{Command}";
+        public void Execute(IConsoleShell shell, string argStr, string[] args)
+        {
+            if(shell.Player != null)
+                EntitySystem.Get<GhostRoleSystem>().OpenEui((IPlayerSession)shell.Player);
+            else
+                shell.WriteLine("You can only open the ghost roles UI on a client.", Color.Red);
```
Comment (metalgearsloth) [inline Content.Server/GameObjects/EntitySystems/GhostRoleSystem.cs:141]:
```suggestion
                shell.WriteLine("You can only open the ghost roles UI on a client.");
```

**Review by metalgearsloth — CHANGES_REQUESTED** [review]
nullables. I also need to suss it ingame but I'll do that later.

File: Content.Server/GameObjects/EntitySystems/GhostRoleSystem.cs (hunk 4 above)
Comment (Zumorica) [reply Content.Server/GameObjects/EntitySystems/GhostRoleSystem.cs:52]: As discussed on discord, this should be fine as all references to the EUI are removed when it's closed.

Comment (Zumorica) [reply Content.Server/GameObjects/EntitySystems/GhostRoleSystem.cs:52]: As discussed on discord, this should be fine as all references to the EUI are removed when it's closed.

Comment (metalgearsloth) [reply Content.Server/GameObjects/EntitySystems/GhostRoleSystem.cs:52]: As discussed on discord, this should be fine as all references to the EUI are removed when it's closed.

**Review by metalgearsloth — APPROVED** [review]

**Pull request merged**